### PR TITLE
More coverage

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Refactor Montgomery arithmetic APIs in strict, basic, and constrained modules to return Option<T> for error handling instead of panicking, propagate failures through the call chain, and expand test coverage with numerous cases for failure paths and edge conditions

Enhancements:
- Change Montgomery N' computation functions to return Option<T> instead of panicking on failure
- Update Montgomery parameter computation and mod mul/exp functions across strict, basic, and constrained modules to propagate errors via Option
- Replace panic branches in trial search, extended Euclidean, and Hensel's lifting with None results and propagate using '?'

Tests:
- Unwrap successful calls in existing Montgomery tests and add unwraps in test modules
- Add extensive tests for Montgomery modules covering invalid inputs, parameter computation failures, and edge-case branches for strict, basic, and constrained variants
- Add new tests in signed.rs for Signed<T> From/into_inner, arithmetic operations, and partial comparisons

## Summary by CodeRabbit

* **New Features**
  * Improved error handling for Montgomery arithmetic operations by returning `None` on invalid inputs instead of panicking.
  * Montgomery parameter computation, multiplication, and exponentiation functions now explicitly indicate failure cases.
  * Enhanced documentation and usage clarity for signed integer arithmetic with detailed examples.
  * Added extensive tests covering signed integer conversions, addition with sign flips, multiplication, and partial ordering edge cases.

* **Bug Fixes**
  * Corrected addition assignment for signed integers to properly handle sign flips and zero canonicalization.
  * Functions now gracefully handle invalid inputs, preventing crashes and ensuring clear error signaling.

* **Tests**
  * Added comprehensive unit tests for Montgomery arithmetic and signed integer operations, including edge cases and failure scenarios.
  * Updated tests to explicitly handle optional return types and verify error handling across all numeric types.

### CODE REVIEW

File: modmath/src/inv/signed.rs [hunk 1]
```diff
@@ -297,4 +297,291 @@ mod signed_tests {
...
+
+    #[test]
+    #[should_panic(expected = "attempt to subtract with overflow")]
+    fn test_signed_add_assign_underflow_bug() {
+        // This test documents a bug in the current AddAssign implementation
+        // When adding to a negative number causes underflow, it panics instead of handling sign flip
+        let mut a = Signed::new(8u32, true); // -8
+        a += &12u32; // This should be -8 + 12 = +4, but current impl tries 8 - 12 and panics
+    }
+
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/inv/signed.rs]: doublecheck this - i think it's fixed

File: modmath/src/inv/signed.rs [hunk 2]
```diff
@@ -297,4 +297,291 @@ mod signed_tests {
...
+    #[test]
+    #[should_panic(expected = "attempt to subtract with overflow")]
+    fn test_signed_add_assign_sign_flip_negative_to_positive_bug() {
+        // Test case that should flip from negative to positive but currently panics
+        // -3 + 7 should equal +4
+        let mut a = Signed::new(3u32, true); // -3
+        a += &7u32; // Should become +4, but panics due to 3 - 7 underflow
+    }
+
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/inv/signed.rs]: doublecheck this - i think it's fixed

File: modmath/src/inv/signed.rs [hunk 3]
```diff
@@ -297,4 +297,291 @@ mod signed_tests {
...
+    #[test]
+    #[should_panic(expected = "attempt to subtract with overflow")]
+    fn test_signed_add_assign_small_negative_large_positive_bug() {
+        // Test case: small negative + large positive should become positive
+        // -1 + 10 should equal +9
+        let mut a = Signed::new(1u32, true); // -1
+        a += &10u32; // Should become +9, but panics due to 1 - 10 underflow
+    }
+
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/inv/signed.rs]: doublecheck this - i think it's fixed

File: modmath/src/inv/signed.rs [hunk 4]
```diff
@@ -297,4 +297,291 @@ mod signed_tests {
...
+        let result = a * &3u32; // +8 * 3 = +24
+        assert_eq!(result.value, 24u32);
+        assert_eq!(result.negative, false);
+
+        // Test Mul<&T> - negative Signed * &T
+        let a = Signed::new(9u32, true); // -9
+        let result = a * &2u32; // -9 * 2 = -18
+        assert_eq!(result.value, 18u32);
+        assert_eq!(result.negative, true);
+    }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/inv/signed.rs:664]: doublecheck this - i think it's fixed
